### PR TITLE
Implemented pagination for facilityuser

### LINKED ISSUE #8221: Don't load all users for a facility at once in facility admin
## Observed behavior
As the number of users in a facility increases the time taken to load them all into the facility user administration page increases. When it gets to a very large number, this can take some considerable time.

## Expected behavior
Kolibri should paginate the users returned to the frontend.

In addition, this would require that the user search functionality be reimplemented, in order to get results from the backend, rather than doing filtering in the client side.

## User-facing consequences
Can lead to delays to delays loading the user page with large numbers of users (reports at ~30,000 users).
…

## Steps to reproduce
Make a large number (~30,000) users in a facility and load the Facility Admin user management page.

### PR DESCRIPTION
**Pull request by navdeepsingh92**
## Summary
Added pagination in facilityuser api to decrease the load time for large amount of data and improve the system performance.
* Backend changes
   1. pagination added in facilityuser api
   2. filter added for usert_type , exclude_member_of and exclude_user_type

* Frontend changes
  - Page (Facility -> users )
    1. implemented pagination with backend api
    2. implemented searching with backend api
    3. implemented user type filtering with backend api
  - Page (Facility -> classes -> enroll learners)
    1. return users from api which are not in current class rather than fetching all and then calculating on frontend.
    2. implemented pagination with backend api
    3. implemented searching with backend api
  - Page (Facility -> classes -> assign coaches)
    1. implemented pagination with api
    2. implemented searching with api

Fixes #8221 

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

**Comment by rtibbles** [comment]
Thanks for the contribution here - some follow up to do here before we can fully review and merge.

1. Please look at the checks for the pull request and make sure all issues are addressed - I would recommend using pre-commit locally to ensure that linting happens before you commit - see our developer documentation for more info: https://kolibri-dev.readthedocs.io/en/develop/getting_started.html#linting-and-auto-formatting
2. You have accidentally committed some files here, because you were developing without git lfs installed. We cannot merge these commits as is, and they will need to be rewritten to exclude these binary files, as otherwise it will pollute the git history of the repo. I would recommend installing git lfs and then doing an interactive rebase to remove these files from the commit history.

**Comment by navdeepsingh92** [comment]
thanks @rtibbles for your response. i will look into these issues.

**Comment by navdeepsingh92** [comment]
Hey @rtibbles . have a look.

File: kolibri/core/assets/src/utils/filterUsersByNames.js
Comment (jredrejo) [inline kolibri/core/assets/src/utils/filterUsersByNames.js:2]: why did you modify this file?

File: kolibri/core/assets/src/views/PaginatedListContainer.vue
Comment (jredrejo) [inline kolibri/core/assets/src/views/PaginatedListContainer.vue]: this indentation change does not follow our [frontend code conventions](https://kolibri-dev.readthedocs.io/en/develop/frontend_architecture/conventions.html)

File: kolibri/core/assets/src/views/PaginatedListContainer.vue
Comment (jredrejo) [inline kolibri/core/assets/src/views/PaginatedListContainer.vue]: why did you remove the clamping that fixes the IE and edge issue? can't happen anymore?

File: kolibri/core/auth/api.py
Comment (jredrejo) [inline kolibri/core/auth/api.py]: please, remove the # added comment. It's not needed with git

File: kolibri/plugins/coach/class_summary_api.py
Comment (jredrejo) [inline kolibri/plugins/coach/class_summary_api.py]: I guess you don't want these comments in the final code

File: kolibri/plugins/facility/assets/src/modules/userManagement/handlers.js
Comment (jredrejo) [inline kolibri/plugins/facility/assets/src/modules/userManagement/handlers.js]: please remove unused code

File: kolibri/plugins/facility/assets/src/modules/userManagement/handlers.js
Comment (jredrejo) [inline kolibri/plugins/facility/assets/src/modules/userManagement/handlers.js]: debugging code left

File: kolibri/plugins/facility/assets/src/views/ClassEnrollForm.vue
Comment (jredrejo) [inline kolibri/plugins/facility/assets/src/views/ClassEnrollForm.vue]: please don't left unused code

**Review by jredrejo — DISMISSED** [review]
hi @navdeepsingh92 I have done a quick check of the code (not functionality  yet) and it's not acceptable yet. Please, as @rtibbles previously advised install the commit hooks as explained at https://kolibri-dev.readthedocs.io/en/develop/getting_started.html#pre-commit-hooks You'll save a lot of time and problems with the tests (that are not passing as you can check)

**Comment by navdeepsingh92** [comment]
Hey @jredrejo. Thanks for review. I have used pre-commit as suggested.

**Comment by rtibbles** [comment]
Will need to update the test suite here: `kolibri/plugins/facility/assets/src/views/UserPage/__test__/UserPage.spec.js` to reflect the  updates here.

**Comment by rtibbles** [comment]
Any updates here?

**Comment by navdeepsingh92** [comment]
Yes, I am working on it. Will update here ASAP.

**Comment by navdeepsingh92** [comment]
Hey, @rtibbles. I have made the changes in the code to pass all the test cases. I have checked on my local and pushed the changes. Kindly review, thanks

**Comment by navdeepsingh92** [comment]
Hey @rtibbles,
Thanks for your feedback and running the checks for merging, we are left with one check of Docs / Checking docs failed. Can you please help us with this? As I checked other PR also has the same check failed.

**Comment by rtibbles** [comment]
Hi @navdeepsingh92 - yes, the docs build failure has just been fixed on release-v0.15.x, so not to worry!

File: kolibri/core/auth/api.py
Comment (rtibbles) [inline kolibri/core/auth/api.py]: Could you rename this variable to `USER_TYPE_CHOICES` to be more explicit about what it is for?

File: kolibri/core/auth/api.py
Comment (rtibbles) [inline kolibri/core/auth/api.py]: I think it would be preferable to do an early return here with this:
```suggestion
            return queryset.filter(roles__isnull=True)
```

File: kolibri/core/auth/api.py
Comment (rtibbles) [inline kolibri/core/auth/api.py]: Likewise here:
```suggestion
            return queryset.exclude(roles__isnull=True)
```

File: kolibri/core/auth/api.py
Comment (rtibbles) [inline kolibri/core/auth/api.py]: The exclude filters are not listed here, are they actually being used? If not, please delete the methods and fields above to limit cruft.

File: kolibri/core/content/utils/annotation.py
Comment (rtibbles) [inline kolibri/core/content/utils/annotation.py]: Thanks - this did not used to get flagged by our linter, but is now. The syntax is required for SQLAlchemy.

File: kolibri/plugins/coach/class_summary_api.py
Comment (rtibbles) [inline kolibri/plugins/coach/class_summary_api.py:284]: Should revert this change as it's not being used anywhere in the frontend.

File: kolibri/plugins/facility/assets/src/modules/userManagement/handlers.js
Comment (rtibbles) [inline kolibri/plugins/facility/assets/src/modules/userManagement/handlers.js:17]: What's the reason for removing this mapper?

File: kolibri/core/assets/src/views/PaginatedListContainer.vue
Comment (rtibbles) [inline kolibri/core/assets/src/views/PaginatedListContainer.vue]: Please leave the code comment here for future reference.

File: kolibri/core/assets/src/views/PaginatedListContainer.vue
Comment (rtibbles) [inline kolibri/core/assets/src/views/PaginatedListContainer.vue]: We cannot change strings in a patch release, please revert this.

File: kolibri/core/assets/src/views/PaginatedListContainer.vue
Comment (rtibbles) [inline kolibri/core/assets/src/views/PaginatedListContainer.vue]: I would prefer to keep this interface of emitting page changes and deferring data loading to the place that is using the `PaginatedListContainer`. At the moment data loading is happening in both the `kolibri/plugins/facility/assets/src/modules/classAssignMembers/handlers.js` file and in here.

The other issue is that this will now break the other uses of this component.

I would suggest adding a prop to the component called `allItems` that indicates whether all the items have been passed (as is the current usage) or whether it is a pre-paginated subset, as you are now implementing. We can't merge this in its current state as it will break functionality in multiple other parts of the app.

**Review by rtibbles — CHANGES_REQUESTED** [review]
Need to defer new functionality into the facility plugin UserPage, rather than breaking the API of the shared PaginatedListContainer component. Can enhance the API to accommodate this use case, but the old API must continue to work.

File: kolibri/core/assets/src/views/PaginatedListContainer.vue
Comment (navdeepsingh92) [reply kolibri/core/assets/src/views/PaginatedListContainer.vue]: Thanks, @rtibbles for your review. I will look into these issues and make the required changes.

Comment (navdeepsingh92) [reply kolibri/core/assets/src/views/PaginatedListContainer.vue]: I think we can use prop `isPaginated` to check whether we have to run the old code or the new one. So we have to make the change in `UserPage` component only and it will have no effect on other pages.

**Comment by navdeepsingh92** [comment]
Hey @rtibbles, I have created a new component `PaginatedListContainerWithBackend` with backend pagination and separated it from `PaginatedListContainer`. Can you have a look and let me know if this will work fine.

**Review by rtibbles — COMMENTED** [review]
Feels a lot cleaner with the separate implementation, thanks - reduces concern about any regressions.

**Comment by navdeepsingh92** [comment]
Thanks, @rtibbles, Let me know if I need to make any other changes.

**Comment by radinamatic** [comment]
Tested the EXE asset on Windows 10 with both Firefox and Chrome, and found several issues.

Working as described by current scenarios:
1. Pagination on *Facility > Users* ✔️ 
2. Pagination on *Facility > Classes > Enroll learners* ✔️ 
3. Pagination on *Coach > Plan > Quiz/Lesson > Recipients > Individual learners* ✔️ 

**Issues**:

1. Right arrow on paginated list in *Facility > Classes > Assign coaches* is **not clickable**. 🔴 

   assign-coach

2. *Create new user* is not working from UI (I had to use the *Import CSV* feature to create users for the above testing scenarios). 🔴 As soon as the focus is in any of the input fields, ```TypeError: this.facilityUsers.find is not a function``` error is visible in the console, and the user can't save the credentials inputed in the form.

   https://user-images.githubusercontent.com/1457929/157350727-801a058a-10b8-45a0-a54b-d02216d6f62f.mp4

**Questions and concerns**:

1. Pagination on *Coach > Plan > Quiz/Lesson > Recipients > Individual learners* has the cutoff at 50 instead of 30 users. ⁉️ 

   https://user-images.githubusercontent.com/1457929/157350960-3a67888c-994a-4c9d-8c9c-725326030240.mp4

 2. Pagination is implemented on *Facility > Users*, but if *Facility > Classes > 'Class'* has more than 30 users, they are all displayed in one long page, **without pagination**. ⁉️ 

   https://user-images.githubusercontent.com/1457929/157351322-cc1263e2-2e6d-4b4d-8779-34429111fcde.mp4

 3. Is this PR up to date with the latest on 0.15 😕? There is the old **Library** link on _Learn > Home_, and _Learn > Bookmarks_, and the _Learn > Home_ is displaying an empty page with the loader, for learners who are not enrolled in any class.

   https://user-images.githubusercontent.com/1457929/157352138-114e7319-2e1e-4051-8800-860fa741d2f4.mp4

**Comment by radinamatic** [comment]
@pcenov Please re-test on Linux to confirm or offer more info. Thank you!

**Comment by pcenov** [comment]
Hi @radinamatic I confirm the issues reported by you are also extant in the Linux build plus I found the following issues:
1. Facility > Users - Attempting to delete a user results in the modal window remaining visible at the screen with disabled Delete and Cancel buttons so the user has to manually reload the page:



2. Class > Enroll learners - If I enroll some users and then click again the Enroll learners button I still see the already enrolled users displayed on the page while the expected behavior is to show only the learners that are not enrolled in the class. If I go ahead and attempt to reenroll them I get a Bad request error in the console:

**Comment by navdeepsingh92** [comment]
Thanks, @radinamatic and @pcenov for your review. I will look into these issues. And I have implemented backend pagination only at the above-mentioned pages. It is not implemented on quiz recipients and the class page.  Let me know if we need this at there as well.

**Comment by navdeepsingh92** [comment]
Hey, @radinamatic @pcenov I have fixed all the above-mentioned issues. And @radinamatic talking about your second question currently this how users are shown [here](https://github.com/learningequality/kolibri/blob/release-v0.15.x/kolibri/plugins/facility/assets/src/views/UserTable.vue#L60). We are not using any paginated container here. Let me know if we also need to use pagination here.

**Comment by pcenov** [comment]
@radinamatic I've retested the issues reported by you and me above and can confirm that these are fixed now. I've used the Linux build as the Windows one is failing here: https://buildkite.com/learningequality/kolibri-python-package/builds/5096
@navdeepsingh92 there's still one issue that needs addressing. Here are the steps to replicate it:
1. Go to Facility>Users
2. Click the pager arrows to go to the last available page and delete all users. Deleting the last available user on the page results in an error.

https://user-images.githubusercontent.com/79847249/158408140-5ce572e9-fdb5-4435-8078-5020dabaeee3.mp4

**Comment by navdeepsingh92** [comment]
Hey, @pcenov I have fixed this issue. you can test.

**Comment by pcenov** [comment]
Thanks @navdeepsingh92, I retested everything today and the last remaining issue about the deleted user is fixed.
I can still see some 404 errors in the console after deleting the last remaining user on the page, are those a matter of concern?
Screenshot:

**Comment by rtibbles** [comment]
It seems like it is making a request with empty query parameters that is giving that 404, so we should definitely clean that up.

**Comment by navdeepsingh92** [comment]
Hi, @rtibbles @pcenov. let me tell you the reason behind that 404 request. Let's suppose you are on page no. 5. and there is only one user on that page. So, when you delete that last user then the frontend makes an API call to get users with page no. 5 but now since there is not any user that can be returned on page 5 that's why Django returns a 404 response. and then the frontend again tries to get users with page no 4 and then everything works fine. That API call will not cause any issue. But let me know if you think we should prevent that call. Also, I have not changed the code for deleting users so that it doesn't affect other parts of the code.

**Comment by navdeepsingh92** [comment]
Hey @rtibbles, please let me know If there is anything else that needs to be fixed.

**Comment by radinamatic** [comment]
Docs build fail is unrelated, but the lint is failing too for ``black``... Can you run  the ``pre-commit run --all-files`` command?

**Comment by navdeepsingh92** [comment]
Thanks, @radinamatic I tried running `pre-commit run --all-files`  and everything was fine.

image (125)

**Comment by radinamatic** [comment]
And if you run the ``git status`` now, are there changes in any of the files that you should commit?

**Comment by navdeepsingh92** [comment]
Nope

**Comment by navdeepsingh92** [comment]
This is a recent commit from `black` https://github.com/psf/black/commit/a66016cb949590863d11e329e65bbb383727125e. maybe this line is not getting ignored.

**Comment by rtibbles** [comment]
I have made an update to the release-v0.15.x that fixes the issue you are seeing with linting - if you rebase your branch and force push it here, it should be fixed.

**Comment by navdeepsingh92** [comment]
Thanks, @rtibbles let me check.

**Comment by navdeepsingh92** [comment]
Hey, @rtibbles I have rebased the branch with release-v0.15.x. I think now it should be fixed.

**Comment by rtibbles** [comment]
Looks like PyPi is having an outage now - so tests are failing as a result.

**Comment by navdeepsingh92** [comment]
@rtibbles can you run tests again?

File: kolibri/plugins/coach/class_summary_api.py
Comment (rtibbles) [reply kolibri/plugins/coach/class_summary_api.py:284]: This change still needs to be reverted.

File: kolibri/core/assets/src/views/PaginatedListContainerWithBackend.vue
Comment (rtibbles) [inline kolibri/core/assets/src/views/PaginatedListContainerWithBackend.vue]: This `_userState` mapper is making this component specific to display in the facility plugin context. As such, it should be removed from the core API and moved into the facility plugin specifically.

This makes even more sense when you consider that anything loaded into the coreAPI is loaded on every single page in Kolibri, but this is only used in the facility plugin.

File: kolibri/core/assets/src/utils/filterUsersByNames.js
Comment (rtibbles) [reply kolibri/core/assets/src/utils/filterUsersByNames.js:2]: This change should be reverted.

**Review by rtibbles — CHANGES_REQUESTED** [review]
Code is nearly there, but a small amount of cleanup is needed here.

**Comment by navdeepsingh92** [comment]
Thanks, @rtibbles I have moved this component to the facility plugin. So now I think it makes more sense because it is only used in the facility plugin.

**Review by rtibbles — APPROVED** [review]
Changes implemented, QA passes. This should be released in 0.15.3.

**Pull request merged**
